cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/VerificationStage.java

```diff
@@ -173,6 +155,66 @@ public void process(ApplicationDeployable input) throws Exception {
     emit(input);
   }
 
+  private void verifyWorkflowSpecifications(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec) {
+    List<WorkflowNode> nodes = workflowSpec.getNodes();
+    for (WorkflowNode node : nodes) {
+      verifyWorkflowNode(appSpec, workflowSpec, node);
+    }
+  }
+
+  private void verifyWorkflowNode(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
+                                  WorkflowNode node) {
+    WorkflowNodeType nodeType = node.getType();
+    switch (nodeType) {
+      case ACTION:
+        WorkflowActionNode actionNode = (WorkflowActionNode) node;
+        verifyWorkflowAction(appSpec, actionNode.getProgram());
+        break;
+      case FORK:
+        WorkflowForkNode forkNode = (WorkflowForkNode) node;
+        if (forkNode.getBranches().size() == 0) {
```

Comment:
This should be a warning in IntelliJ. You should use `isEmpty()` instead of `size() == 0`.